src/edu/csus/ecs/pc2/clics/API202306/CLICSTeam.java

```diff
@@ -0,0 +1,246 @@
+        }
+    }
+
+    /**
+     * Create account list from a teams.json like file
+     *
+     * @param contest the contest (needed for groups)
+     * @param jsonfile json file to deserialize
+     * @param site the site to create the accounts for
+     * @param institutionsMap hashmap mapping organization id to organization fields (to be removed once organizations
+     *     are part of the contest model like groups). May be null if not interested in organizations.
+     * @return array of accounts to add, or null on error
+     */
+    public static Account [] fromJSON(IInternalContest contest, File jsonfile, int site, HashMap<String,String[]> institutionsMap) {
+        Account [] newaccounts = null;
+        Log log = StaticLog.getLog();
+
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            newaccounts = createTeamsFromTeamsJSON(contest, mapper.readValue(jsonfile, CLICSTeam[].class), site, institutionsMap, log);
+        } catch (Exception e) {
+            log.log(Log.WARNING, "could not deserialize team file " + jsonfile, e);
+        }
+        return(newaccounts);
+     }
+
+    /**
+     * Create account list from a strings contains a teams json file
+     *
+     * @param contest the contest (needed for groups)
+     * @param json json to deserialize
+     * @param site the site to create the accounts for
+     * @param institutionsMap hashmap mapping organization id to organization fields (to be removed once organizations
+     *     are part of the contest model like groups).  May be null if not interested in organizations.
+     * @return array of accounts to add, or null on error
+     */
+    public static Account [] fromJSON(IInternalContest contest, String json, int site, HashMap<String,String[]> institutionsMap) {
+        Account [] newaccounts = null;
+        Log log = StaticLog.getLog();
+
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            newaccounts = createTeamsFromTeamsJSON(contest, mapper.readValue(json, CLICSTeam[].class), site, institutionsMap, log);
+        } catch (Exception e) {
+            log.log(Log.WARNING, "could not deserialize team string", e);
+        }
+        return(newaccounts);
+     }
+
+    /**
+     * Converts CLICS teams into a PC2 Account array.
+     *
+     * @param contest The contest
+     * @param teams array of CLICS teams
+     * @param site The site to create accounts on
+     * @param log For errors
+     * @return an Account array of teams converted from the CLICS teams or null if error
+     */
+    private static Account [] createTeamsFromTeamsJSON(IInternalContest contest, CLICSTeam [] teams, int site, HashMap<String,String[]> institutionsMap, Log log) {
+
+        Account [] newaccounts = null;
+        ArrayList<Account> accounts = new ArrayList<Account>();
+        Account account;
+        JSONTool jsontool = new JSONTool(contest, null);
```

Comment:
There are multiple methods within class `JSONTool` that use `controller` (whose value is passed in here as `null`).  I realize that none of the `JSONTool` methods currently being called in this class use the `controller`, but what about future updates?  Someone could add code this class to call a different `JSONTool` method, not realizing that at runtime this could generate an error (and such error might not show up for a long time).  For safety it seems like `JSONTool` should be constructed with a `controller`.  Is there any reason not to require passing a `controller` into the `fromJSON()` methods (so that they can forward that controller to this method and then it can be passed to the `JSONTool` constructor?

Reply:
IIRC it may have had to do with test units not having a controller handy, or, since this is used at contest load time, it may have to do with not have a controller when the JSON is being read in (we haven't written the code to read the contest configuration from JSON files - yet).  Also, it's only used for logging, and dutifully checked for being `null `before being dereferenced.